pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/AvgShedder.java

```diff
@@ -0,0 +1,320 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.loadbalance.impl;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.hash.Hashing;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Random;
+import java.util.Set;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.mutable.MutableDouble;
+import org.apache.commons.lang3.mutable.MutableInt;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.pulsar.broker.ServiceConfiguration;
+import org.apache.pulsar.broker.loadbalance.LoadData;
+import org.apache.pulsar.broker.loadbalance.LoadSheddingStrategy;
+import org.apache.pulsar.broker.loadbalance.ModularLoadManagerStrategy;
+import org.apache.pulsar.policies.data.loadbalancer.BrokerData;
+import org.apache.pulsar.policies.data.loadbalancer.BundleData;
+import org.apache.pulsar.policies.data.loadbalancer.LocalBrokerData;
+import org.apache.pulsar.policies.data.loadbalancer.TimeAverageMessageData;
+
+@Slf4j
+public class AvgShedder implements LoadSheddingStrategy, ModularLoadManagerStrategy {
+    // map bundle to broker.
+    private final Map<BundleData, String> bundleBrokerMap = new HashMap<>();
+    // map broker to Scores. scores:0-100
+    private final Map<String, Double> brokerScoreMap = new HashMap<>();
+    // map broker hit count for high threshold/low threshold
+    private final Map<String, MutableInt> brokerHitCountForHigh = new HashMap<>();
+    private final Map<String, MutableInt> brokerHitCountForLow = new HashMap<>();
+
+    // result returned by shedding, map broker to bundles.
+    private final Multimap<String, String> selectedBundlesCache = ArrayListMultimap.create();
+
+    private static final double MB = 1024 * 1024;
+    private static final Random random = new Random();
+
+    @Override
+    public Multimap<String, String> findBundlesForUnloading(LoadData loadData, ServiceConfiguration conf) {
+        selectedBundlesCache.clear();
```

Comment:
Should it be a local variable? I see it's only used in `findBundlesForUnloading` and `selectBundleForUnloading` (also called by `findBundlesForUnloading`)

Reply:
Let it globally could avoid frequent GC. Same implementation as `ThresholdShedder`.
```
public class ThresholdShedder implements LoadSheddingStrategy {
    private static final Logger log = LoggerFactory.getLogger(ThresholdShedder.class);
    private final Multimap<String, String> selectedBundlesCache = ArrayListMultimap.create();
```

Reply:
The implementation of `ThresholdShedder` is bad from my perspective.

Making it a class field only reduces the construction and destruction of a `Multimap` object. But the main object allocation and deallocation should come from entries in the map. Extending the lifetime of such maps is meaningless here unless you can prove it could bring GC pressure.

Let's avoid premature optimization.

Reply:
Make sense, i update the implementation, thanks for review.